// Swappers/SWO.js
module.exports = SWO;

var data = require('FileData').get(),
    grader = require('Grade');

var maxGroupSize;

function SWO() {
    var results = []
    var students = data.student.students
    var actualPossibleTimes = []
    var groups = data.class.groups
    maxGroupSize = data.objective.groupSize.max
    console.log("max is: " + maxGroupSize)
    var studentsArr = []
    for (var i in students) studentsArr.push(students[i])
    for (var i = 0; i<10; i++){
        var assignment = {}
        for (var groupTime in groups) {
            assignment[groupTime] = []
            console.log(groupTime)
            actualPossibleTimes.push(groupTime)
        }
        for (var j in studentsArr) {
            var time = getTimeForStudent(studentsArr[j], actualPossibleTimes, assignment)
            assignment[time].push(studentsArr[j].email)
            studentsArr[j].assignment = time
            studentsArr[j].satisfaction = rateAssignment(studentsArr[j])
            console.log(studentsArr[j].satisfaction)
        }
        console.log("----------------------------------------------------------------------")
        var screwed = studentWhoGotScrewed(studentsArr)
        for (var k in screwed) console.log(screwed[k].email)
        console.log("----------------------------------------------------------------------")
        var withScrewedRemoved = removeStudents(studentsArr, screwed)
        studentsArr = screwed.concat(withScrewedRemoved)
        for (var l in studentsArr) console.log(studentsArr[l].email)
        results.push(assignment)
    }
}

function getTimeForStudent(student, possibleTimes, assignment){
    for(var i in student.goodTimes){
        for(var j in possibleTimes){
            if(student.goodTimes[i] == possibleTimes[j] && isOption(possibleTimes[i],assignment)) return possibleTimes[j]
        }
    }
    for(var i in student.possibleTimes){
        for(var j in possibleTimes){
            if(student.possibleTimes[i] == possibleTimes[j] && isOption(possibleTimes[i],assignment)) return possibleTimes [j]
        }
    }
    var time;
    var done = false;
    while(!done){
        time = possibleTimes[randNumber(0,possibleTimes.length)]
        if(isOption(time,assignment)){
            break;
        }
    }
    return time;
}

function rateAssignment(student){
    if(student.goodTimes.indexOf(student.assignment) > -1) return 0;
    if(student.possibleTimes.indexOf(student.assignment) > -1) return 1;
    return 2;
}

function isOption(time, assignment){
    if(assignment[time].length < maxGroupSize) return true;
    return false;
}

function randNumber(min,max){
    return Math.floor(Math.random() * (max-min) + min);
}

function studentWhoGotScrewed(students){
    var copyOfStudents = students.slice();
    copyOfStudents.sort(function(a,b){
        return a.satisfaction - b.satisfaction;
    })
    return copyOfStudents.slice(0,5)
}

function removeStudents(students, toBeRemoved){
    var retArray = []
    for(var i in students){
        var wasthere = false;
        for(var j in toBeRemoved){
            if(students[i].email == toBeRemoved[j].email) wasthere=true;
        }
        if(!wasthere) retArray.push(students[i])
    }
    return retArray
}
